python/ray/dag/tests/experimental/test_torch_tensor_dag.py

```diff
@@ -863,6 +909,1086 @@ def test_torch_tensor_exceptions(ray_start_regular):
+
+    nccl_group = AbstractNcclGroup([workers[0]])
+    with InputNode() as inp:
+        computes = [
+            worker.compute_with_tuple_args.bind(inp, i)
+            for i, worker in enumerate(workers)
+        ]
+        with pytest.raises(
+            ValueError,
+            match="Expected actor handles to match the custom NCCL group",
+        ):
+            collective.allreduce.bind(computes, transport=nccl_group)
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_nccl_comm_deduplicate_collectives(ray_start_regular, monkeypatch):
+    """
+    Test communicators are deduplicated when all-reduce is called on the same
+    group of actors more than once.
+    """
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
+
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 1
+    ), "This test requires at least 2 GPUs"
+
+    # Mock the `_init_nccl_group` function to keep track of the NCCL group IDs.
+    init_nccl_group_set = InitNcclGroupSet()
+    monkeypatch.setattr(
+        "ray.dag.collective_node._init_nccl_group",
+        init_nccl_group_set,
+    )
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+
+    num_workers = 2
+    workers = [actor_cls.remote() for _ in range(num_workers)]
+
+    with InputNode() as inp:
+        computes = [
+            worker.compute_with_tuple_args.bind(inp, i)
+            for i, worker in enumerate(workers)
+        ]
+        collectives = collective.allreduce.bind(computes)
+        collectives = collective.allreduce.bind(collectives)
+        recvs = [
+            worker.recv.bind(collective)
+            for worker, collective in zip(workers, collectives)
+        ]
+        dag = MultiOutputNode(recvs)
+
+    compiled_dag = dag.experimental_compile()
+
+    from ray.experimental.channel import ChannelContext
+
+    # Only 1 NCCL group should be created.
+    assert len(init_nccl_group_set.nccl_group_ids) == 1
+    nccl_group_id = list(init_nccl_group_set.nccl_group_ids)[0]
+    ctx = ChannelContext.get_current()
+    nccl_group = ctx.nccl_groups[nccl_group_id]
+    nccl_actors = nccl_group.get_actor_handles()
+    # The NCCL group should contain both workers.
+    assert set(nccl_actors) == set(workers)
+    # P2P NCCL group should be none since NCCL transport is not used.
+    assert compiled_dag._nccl_group_id_p2p is None
+
+    # Sanity check: the compiled dag can execute.
+    shape = (10,)
+    dtype = torch.float16
+    ref = compiled_dag.execute([(shape, dtype, i + 1) for i in range(num_workers)])
+    result = ray.get(ref)
+    reduced_val = sum(i + 1 for i in range(num_workers)) * 2
+    assert result == [(reduced_val, shape, dtype) for _ in workers]
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_nccl_comm_deduplicate_p2p_and_collective(
```

Comment:
These tests can be significantly simplified if you break out the NCCL group initialization code into a helper function that can be unit-tested on its own without any dependency on NCCL or a custom communicator group.

We can keep the monkeypatched tests, but just for the cases where we test e2e whether the right groups are initialized and destroyed.